ssl/record/rec_layer_s3.c

```diff
@@ -1554,30 +1554,44 @@ int ssl3_read_bytes(SSL *s, int type, int *recvd_type, unsigned char *buf,
         return -1;
     }
 
-    /*
-     * If we've sent a close_notify but not yet received one back then ditch
-     * anything we read.
-     */
     if ((s->shutdown & SSL_SENT_SHUTDOWN) != 0) {
-        /*
-         * In TLSv1.3 this could get problematic if we receive a KeyUpdate
-         * message after we sent a close_notify because we're about to ditch it,
-         * so we won't be able to read a close_notify sent afterwards! We don't
-         * support that.
-         */
-        SSL3_RECORD_set_length(rr, 0);
-        SSL3_RECORD_set_read(rr);
-
         if (SSL3_RECORD_get_type(rr) == SSL3_RT_HANDSHAKE) {
             BIO *rbio;
+            /*
+             * We need to check the message type of the received data. We're not
+             * "in init" otherwise we won't have got here - so this is a new
+             * handshake message. However it's theoretically possible for us to
+             * have received part of the message header before us sending
+             * close_notify, and for us to be receiving the remainder of it now.
+             */
+            unsigned char mt = s->rlayer.handshake_fragment_len > 0
+                               ? s->rlayer.handshake_fragment[0]
+                               : *SSL3_RECORD_get_data(rr);
 
-            if ((s->mode & SSL_MODE_AUTO_RETRY) != 0)
-                goto start;
+            /*
+             * We ignore any handshake messages sent to us unless they are
+             * TLSv1.3 NewSessionTicket or KeyUpdate, in which case we want to
+             * process them. For all other handshake messages we can't do
+             * anything reasonable with them because we are unable to write any
+             * response due to having already sent close_notify.
+             */
+            if (!SSL_IS_TLS13(s)
+                       /* Shouldn't ever be 0 but lets be safe */
+                    || SSL3_RECORD_get_length(rr) == 0
```

Comment:
In case this == 0, isn't the get_data() above wrong?

Reply:
Yes, but Matt claims that this shouldn't ever be zero.  My intuition agrees with Matt, though I am not finding where in the code this is presently enforced.